File: homework3/steven_law/q11_VacationDestinations.cpp
/*Data Structure: Graph (with Priority Queue)
Algorithm: Dijkstra's
Time Complexity: O((V+E) log V)
Space Complexity: O(V+E)
Time Taken: 40 min, partial - stopover logic feels wrong
*/

#include<iostream>
#include<vector>
#include<unordered_map>
#include <string>
#include <unordered_set>
#include <queue>
#include <tuple>

using namespace std;

int vacationDestinations(vector<tuple<string, string, double>> routes, string origin, double k){
    unordered_map<string, vector<pair<string, double>>> graph;

    for(auto& r : routes){
        string a = get<0>(r);
        string b = get<1>(r);
        double t = get<2>(r);
        graph[a].push_back({b, t});
        graph[b].push_back({a, t});
    }

    priority_queue<pair<double, string>, vector<pair<double, string>>, greater<pair<double, string>>> pq;
    unordered_map<string, double> dist;

    pq.push({0.0, origin});
    dist[origin] = 0.0;

    while(!pq.empty()){
        auto [d, city] = pq.top();
        pq.pop();

        if(d > dist[city]) continue;

        for(auto& [neighbor, travelTime] : graph[city]){
            // adding 1 hour for stopover on every edge. not 100% sure if origin should
            // count but going with this for now
            double newDist = d + 1.0 + travelTime;

            if(newDist <= k && (!dist.count(neighbor) || newDist < dist[neighbor])){
                dist[neighbor] = newDist;
                pq.push({newDist, neighbor});
            }
        }
    }

    int count = 0;
    for(auto& [city, d] : dist){
        if(city != origin && d <= k) count++;
    }
    return count;
}

int main(){
    vector<tuple<string, string, double>> routes = {
        {"Boston", "New York", 4},
        {"New York", "Philadelphia", 2},
        {"Boston", "Newport", 1.5},
        {"Washington, D.C.", "Harper's Ferry", 1},
        {"Boston", "Portland", 2.5},
        {"Philadelphia", "Washington, D.C.", 2.5}
    };

    cout << "k=5: " << vacationDestinations(routes, "New York", 5) << endl;
    cout << "k=7: " << vacationDestinations(routes, "New York", 7) << endl;
    cout << "k=8: " << vacationDestinations(routes, "New York", 8) << endl;
}

/* psudocode/thoughts/logic
goal: count destinations within k hours of origin, with 1 hour added per stopover
restraints: stopovers add an hour. unclear if origin counts as a stopover

strategy: ok so first i tried plain bfs but realized that doesnt work since edges have
weights (hours), cant just count hops. switched to dijkstra.

dijkstra is fine for the base case but the stopover thing is what tripped me up. the prompt
says "having a stopover in a city adds an hour" but doesnt clearly say if origin counts.
logically it shouldnt (youre starting there) but i couldnt figure out how to special-case
the origin in dijkstra without making it ugly.

so i just added 1 hour to every edge expansion which i KNOW is wrong since the very first
edge from origin gets a stopover penalty too. that means everything reachable comes out
1 hour further than it should be.

bigger issue too - im pretty sure the way im computing dist isnt counting stopovers right
even on intermediate nodes. dijkstra normally relaxes by edge weight but here the weight
depends on whether youre at the origin or not which makes the relax step non-uniform. that
breaks the optimal substructure property of dijkstra i think??

ran out of time. test 1 might come out right by accident (k=5 small enough that the off-by-1
doesnt change the count) but tests 2 and 3 are def gonna be wrong.

prompt typo note: pdf says "Output: 2" for all three k values but the cities listed match
2/4/6 reachable. pretty sure the "2" is a typo and the actual count is what i listed in
expected.

bringing this one up at the next sync.
*/

File: delight_oti/q11_VacationDestinations.py
from collections import deque

def VacationDestinations(routes, origin, k):
    graph = {}

    for city1, city2, time in routes:
        if city1 not in graph:
            graph[city1] = []
        if city2 not in graph:
            graph[city2] = []

        graph[city1].append((city2, time))
        graph[city2].append((city1, time))

    queue = deque()
    queue.append((origin, 0))

    best_time = {}
    best_time[origin] = 0

    reachable = set()

    while queue:
        city, current_time = queue.popleft()

        for neighbor, travel_time in graph[city]:
            extra_stopover = 0

            if city != origin:
                extra_stopover = 1

            new_time = current_time + extra_stopover + travel_time

            if new_time <= k:
                if neighbor != origin:
                    reachable.add(neighbor)

                    if neighbor not in best_time or new_time < best_time[neighbor]:
                        best_time[neighbor] = new_time
                        queue.append((neighbor, new_time))

    return len(reachable)

routes = [
    ("Boston", "New York", 4),
    ("New York", "Philadelphia", 2),
    ("Boston", "Newport", 1.5),
    ("Washington, D.C.", "Harper's Ferry", 1),
    ("Boston", "Portland", 2.5),
    ("Philadelphia", "Washington, D.C.", 2.5)
]

# origin = "New York"
# k = 5

# origin = "New York"
# k = 7

# origin = "New York"
# k = 8

# print(VacationDestinations(routes, origin, k))

# 40